Source/qol/statistics.cpp

```diff
@@ -0,0 +1,97 @@
+#include "qol/statistics.h"
+
+namespace devilution {
+
+Statistics myPlayerStatistics;
+std::map<std::string, std::string> statisticsFile;
+
+void SaveMonsterKillCount();
+void LoadMonsterKillCount();
+std::string GetMonsterName(uint16_t monsterID);
+void AddMonsterWeapon(std::string &monsterName, uint16_t monsterID);
+
+void InitializePlayerStatistics()
+{
+	myPlayerStatistics = {};
+	std::fill_n(MonsterKillCounts, MAXMONSTERS, 0);
+}
+
+void LoadStatisticsFromMap()
+{
+	if (statisticsFile.find("deathCount") != statisticsFile.end())
+		myPlayerStatistics.deathCount = std::stoul(statisticsFile.at("deathCount"));
+
+	if (statisticsFile.find("ingameTime") != statisticsFile.end())
+		myPlayerStatistics.ingameTime = std::stoul(statisticsFile.at("ingameTime"));
+	LoadMonsterKillCount();
+}
+
+void SaveStatisticsToMap()
+{
+	statisticsFile["deathCount"] = std::to_string(myPlayerStatistics.deathCount);
+	statisticsFile["ingameTime"] = std::to_string(myPlayerStatistics.ingameTime);
+	SaveMonsterKillCount();
+}
+
+void CalculateInGameTime()
+{
+	uint32_t ticksNow = SDL_GetTicks();
+	myPlayerStatistics.ingameTime += (ticksNow - myPlayerStatistics.ticksSubstrahend);
+	myPlayerStatistics.ticksSubstrahend = ticksNow;
+}
+
+void LoadMonsterKillCount()
+{
+	for (uint16_t monsterID = 0; monsterID < _monster_id::NUM_MTYPES; monsterID++) {
+		std::string monsterName = GetMonsterName(monsterID);
+		if (statisticsFile.find("killed." + monsterName) != statisticsFile.end())
+			MonsterKillCounts[monsterID] = std::stoi(statisticsFile.at("killed." + monsterName));
+	}
+}
+
+void SaveMonsterKillCount()
+{
+	for (uint16_t monsterID = 0; monsterID < _monster_id::NUM_MTYPES; monsterID++) {
+		std::string monsterName = GetMonsterName(monsterID);
+		statisticsFile["killed." + monsterName] = std::to_string(MonsterKillCounts[monsterID]);
+	}
+}
+
+std::string GetMonsterName(uint16_t monsterID)
+{
+	std::string monsterName = MonstersData[monsterID].mName;
+	monsterName.erase(std::remove_if(monsterName.begin(), monsterName.end(), [](unsigned char x) { return std::isspace(x); }), monsterName.end());
+	AddMonsterWeapon(monsterName, monsterID);

```

Comment:
This is not taking translations in to consideration, it's probably better have the full name in the switch rather then try to edit it, that way you can also return a string_view instead.

Reply:
Hm this might be for a text based file format, that feel a bit foreign compared to how all other data is saved. And also I'm not sure it's a good idea compared to using there id as the identifier.

Reply:
> This is not taking translations in to consideration, it's probably better have the full name in the switch rather then try to edit it, that way you can also return a string_view instead.

This should not take translations into considerations. I am building statistic row name here (key) for `statisticsFile` map, language independent. Its purpose isn't to display its name. But to be just english internal name for particular statistic row.
currently MonsterKillCounts is an array that stores killed monsters in order based on their ID.
I am taking this ID here, get the name from `MonstersData` correlating to it. And since no all names are unique I am making them unique.
1. First I trim the name from spaces.
2. AddMonsterWeapon function with switch is responsible only to make some non-unique names unique.

i.e. there are in `MonstersData` 2 positions with the same monster name:
- `MT_GGOATMC` `"Night Clan"` and 
- `MT_GGOATBW` `"Night Clan"`
I am making those statistics *key* (in map) name: 
- `killed.NightClan`
- `killed.NightClanArcher`

>Hm this might be for a text based file format, that feel a bit foreign compared to how all other data is saved. And also I'm not sure it's a good idea compared to using there id as the identifier.

yes `statisticsFile` is text base. It is <string, string> map.
keys in there looks like
`killed.NightClan`
`DeathCount`
etc.
and I later save it to file as `"killed.NightClan=666\n"` line - see `SaveStatistics` in loadsave.cpp
I am not using id as identifier at all.

Reply:
Considering that we already have unique ids, and use them as a basis for the generated strings, this seams pointless/wistful.

Reply:
> Considering that we already have unique ids, and use them as a basis for the generated strings, this seams pointless/wistful.

I wanted each key name to be meaningful. Also as follow-up to this PR I thought about adding counting unique monsters killed (like Lazarus) and handle them like that (`killed.BlackJade`). They are not counted separately right now by the game. They are not counted at all. And bosses have no unique identifier... They have only order in ` UniqueMonstersData`. So this would be more consistent to name all monster kills in the same way.

Reply:
Unique also have an index :) I'm a bit unsure about using the display name to generate an index as it was not intended for it, and may be subject to change.

Reply:
why?

Reply:
because BlackJade is not necessarily a monster in a mod? Why not just use the index, so the key is not dependent on what the monster is (or was previously) called?

Reply:
It is not hardcoded. It takes name by index from monstersdat tables. Your mod will have different monster and unique monster names and statistics keyes saved in your mod will be named after the names of your monsters.
What's hardcoded here is only few exceptions in d1 that have several couples of monsters using the same name.

Reply:
I think I misunderstood based on the conversation, since it does seem like it's going to pull the current monster name to generate the key, although the AddMonsterWeapon thing would be applicable to what I was saying - would just need to make sure to update the labels when modding I guess.

Reply:
Actually I think that all literals could be exported to translations. English too ;)
Or alternatively - fix typos only in english translations. But i guess there is no such place. Yet! :grin:

Reply:
Ok I have an idea.
How about change current indentifiers in enum `_monster_id` from like MT_GGOATBW` to a new ones `NightClanArcher` ?

Reply:
Renaming `MT_GGOATBW` and other old names to nice descriptive names like `NightClanArcher` would be much appreciated :)

Reply:
> yes, UniqueMonstersData is an array so yes there's index, but no unique id like normal monsters have and I could use.

Uh didn't even realize that :sweat:

Reply:
> > yes, UniqueMonstersData is an array so yes there's index, but no unique id like normal monsters have and I could use.
> 
> Uh didn't even realize that 😓

We can expand the `UMT_*` enum easily enough.

Reply:
> Renaming `MT_GGOATBW` and other old names to nice descriptive names like `NightClanArcher` would be much appreciated :)

I realized that this still doesn't solve the issue.
There is no clean way of converting enum to string without some convoluted tricks.
This is 2022 and reflection isn't in C++ yet. Its proposal is postponed again to C++23.
You can go here:
https://stackoverflow.com/questions/28828957/enum-to-string-in-modern-c11-c14-c17-and-future-c20
and choose which way would be the best fitting the project.
Amongst many solutions there are prominent:
- enum redeclaration by macro  (or x macro technique)
- using another lib like boost or magic enum or others
- enum generation from external file (even more subject to change)
- just retype all literals again in case switch or other way (which means double typing hardcoded strings).

At this point I would stay with current commited solution and wait for C++23 reflection. But pick one from the link if you really think it is better.

Reply:
I was suggesting not using strings, since we don't do so for any other data type

Reply:
> * just retype all literals again in case switch or other way (which means double typing hardcoded strings).

We already use this strategy for other enums.

https://github.com/diasurgical/devilutionX/blob/39afeaf8b1a474cd662d5975994cb5cb939597f8/Source/dvlnet/packet.cpp#L54-L80

> At this point I would stay with current commited solution and wait for C++23 reflection. But pick one from the link if you really think it is better.

At the very least, we should avoid using `std::isspace()`. This quote is from @glebm on Discord in response to a bug we found recently from using `isblank()` in some parsing code.

> Don't use any of the `is` functions please. They're locale-dependent and basically super-broken because of that

Reply:
> I was suggesting not using strings, since we don't do so for any other data type

@k-bar We can define an enum that maps each statistic to an integer ID and use that as a key for the map. We can also set high bits to create categories for monsters and unique monsters.

What was the purpose for selecting `string` as the value type? Aren't basically all statistics numbers?

Reply:
> What was the purpose for selecting `string` as the value type? Aren't basically all statistics numbers?

Yes, most of the statistics that come to mind are numerical. And most of them will be counters. However, as I mentioned in the description of PR - I wanted the solution to be ready to write also non-numeric values. In addition to the example from the PR description, you can e.g. imagine some last string value (e.g. the name of the game, or the name of the last killed player, whatever).

The decision to use a map with any key in the form of a string was dictated by the need to have the save and read statistics disconnected from the game version.
This way, if a newer version of the game introduces a new statistic, the savegame from that version still works in the previous version, the statistics supported by the previous version are updated, and new, unknown ones are read but unused and not lost. They are stored in the map and saved without loss. The size of the map is also not predetermined in this way.
Such a design seemed to me the safest from data loss (overwriting, unsaved) and flexible in terms of what kind of statistics can be added in the future. It doesn't matter what kind. New keys will just be added to the map.
In case of resignation from the map solution string, string - first of all, the possibility of writing non-string statistics (for value) is given up. Secondly, the more flexible solution ensuring the compatibility of save files (key) is abandoned. (eg how many higher bits should be allocated to types of statistics, what if they run out etc. I assume that what you mentioned would also be a map only with a numeric key, how to save and parse it?). String keys seem like a much simpler solution.

Reply:
> We already use this strategy for other enums.

If this is correct approach for the project, let me know, and I will make proper readjustment.
I am not sure if @DakkJaniels would be happy :)